### ISSUE TITLE
interrupted and/or stopped dscan does not return to initial position (SF#7)

### ISSUE THREAD

**Issue by sf-migrator-bot** [sardana-org/sardana#8]
Note: This is an internal ALBA ticket (RT#21632)that is passed to the sardana tracker.

Sun Jan 29 15:59:26 2012 ffauth - Ticket created 
From: 	F Fauth

Two cases,

1- I launched a dscan but interrupted with ctrlC, is it normal the motor did not go to the initial position
2- I launched a dscan, it started correctly but could not finish because of soft limit. Is it normal the motor did not go to the initial position. ? Is it normal that the scan starts though he will not be able to reach the final position of teh scan ?

Reported by: cpascual ( http://sf.net/u/cpascual )

Original Ticket: [sardana/tickets/7](https://sourceforge.net/p/sardana/tickets/7)

**Comment by reszelaz** [comment]
What was described in my previous comment may also negativelly affect the continuous scans e.g. `ascanc` while being aborted. In this case the `CScan.do_restore` method, that tries to set the original motion paramters like velocity, acceleration, etc., is called in the finally clause of `GScan.step_scan`. Since actual stopping of the motors is being executed in a separate thread, nothing guaratees that they will be already stopped whend doing the restore. When that happens in a scan of a motor which does not allow to change the parameters while being moved an exception is raised leaving the system misconfigured (wrong velocities...). This affects the Icepap controller widely used at Alba:

~~~
SardanaTP.W006 OUTPUT   2017-01-16 12:57:35,282 BL29/DOOR/01.Macro[ascanc]:    16     57.6306      0.05     -0.0024243
2  -0.00201569  -0.00185608  -0.00173462       0.05         799.793          -5489.67      1.48004e+06   169.089
SardanaTP.W006 DEBUG    2017-01-16 12:57:35,286 CSScan: [START] acquisition
Dummy-11       DEBUG    2017-01-16 12:57:37,760 BL29/DOOR/01: stopping macro Macro 'ascanc(xmcd_x, 0.0, 500.0, 0.05, 0
.3) -> 91b10b42-dbe2-11e6-8144-001999bef9d4'
SardanaTP.W006 ERROR    2017-01-16 12:57:37,914 tbl2901:10000.xmcd_x.velocity: [Tango] write failed (PyDs_PythonError)
: IcePAPException: IcePAPException(0,SendWriteCommand failed the 'ACK',Error sending command 2:VELOCITY 120.0, icepap
answered 2:VELOCITY ERROR Motor is moving)

SardanaTP.W006 DEBUG    2017-01-16 12:57:37,914 BL29/DOOR/01.Macro[ascanc]: Failed to restore a backup of the xmcd_x m
otor
SardanaTP.W006 DEBUG    2017-01-16 12:57:37,916 BL29/DOOR/01.Macro[ascanc]: Details:
Traceback (most recent call last):
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/scan/gscan.py", line 1261, in do_restore
    motor.setVelocity(motor_backup['velocity'])
  File "/homelocal/sicilia/lib/python/site-packages/sardana/taurus/core/tango/sardana/pool.py", line 699, in setVeloci
ty
    return self.getVelocityObj().write(value)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/taurus/core/tango/sardana/pool.py", line 237, in write
    self._attr.write(value, with_read=False)
  File "/homelocal/sicilia/lib/python/site-packages/taurus/core/tango/tangoattribute.py", line 245, in write
    raise df
DevFailed: DevFailed[
DevError[
    desc = IcePAPException: IcePAPException(0,SendWriteCommand failed the 'ACK',Error sending command 2:VELOCITY 120.0
, icepap answered 2:VELOCITY ERROR Motor is moving)

  origin =   File "/homelocal/sicilia/lib/python/site-packages/sardana/tango/pool/PoolDevice.py", line 386, in _write_
DynamicAttribute
    return write(attr)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/tango/core/util.py", line 252, in write_attr_wrapper
    ret = write_attr_func(self, attribute)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/tango/pool/Motor.py", line 503, in write_Velocity
    self.motor.velocity = attr.get_write_value()
  File "/homelocal/sicilia/lib/python/site-packages/sardana/pool/poolmotor.py", line 561, in set_velocity
    self.controller.set_axis_par(self.axis, "velocity", velocity)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/pool/poolcontroller.py", line 270, in wrapper
    return fn(pool_ctrl, *args, **kwargs)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/pool/poolcontroller.py", line 495, in set_axis_par
    return self.ctrl.SetAxisPar(axis, name, value)
  File "/homelocal/sicilia/lib/python/site-packages/sardana/pool/controller.py", line 415, in SetAxisPar
    return self.SetPar(axis, parameter, value)
  File "/homelocal/sicilia/lib/python/site-packages/poolcontrollers/IcePAPCtrl/IcePAPCtrl.py", line 432, in SetPar
    self.iPAP.setSpeed(axis, value_steps)
  File "/homelocal/sicilia/lib/python/site-packages/pyIcePAP/IcePAP.py", line 337, in setSpeed
  File "/homelocal/sicilia/lib/python/site-packages/pyIcePAP/IcePAP_Eth.py", line 145, in sendWriteCommand

  reason = PyDs_PythonError
severity = ERR]

DevError[
    desc = Failed to write_attribute on device motor/icepap_icebl2902_ctrl/2, attribute Velocity
  origin = DeviceProxy::write_attribute()
  reason = API_AttributeFailed
severity = ERR]
]
SardanaTP.W006 DEBUG    2017-01-16 12:57:37,918 BL29/DOOR/01.MacroExecutor: Sending stop event
SardanaTP.W006 DEBUG    2017-01-16 12:57:37,920 BL29/DOOR/01.MacroExecutor: [ENDEX] (ScanException) runMacro ascanc
SardanaTP.W006 DEBUG    2017-01-16 12:57:37,921 BL29/DOOR/01: Traceback (most recent call last):
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/msmacromanager.py", line 1224, in runMacro
    for step in macro_obj.exec_():
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/macro.py", line 2265, in exec_
    for i in it:
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/macros/scan.py", line 263, in run
    for step in self._gScan.step_scan():
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/scan/gscan.py", line 865, in step_scan
    self.do_restore()
  File "/homelocal/sicilia/lib/python/site-packages/sardana/macroserver/scan/gscan.py", line 1269, in do_restore
    raise ScanException("error while restoring the motor's backup")
ScanException: error while restoring the motor's backup

SardanaTP.W006 ERROR    2017-01-16 12:57:37,922 BL29/DOOR/01: An error occurred while running Macro 'ascanc(xmcd_x, 0.
0, 500.0, 0.05, 0.3) -> 91b10b42-dbe2-11e6-8144-001999bef9d4':
error while restoring the motor's backup
SardanaTP.W006 DEBUG    2017-01-16 12:57:37,924 BL29/DOOR/01.MacroExecutor: Job ended (stopped=True, aborted=False)
SardanaTP.W007 DEBUG    2017-01-16 12:57:39,371 CSScan: Restored backup of the xmcd_x motor.
~~~

**Comment by reszelaz** [comment]
We need an event flag, let’s use `reserved_objs_stopped` and `reserved_objs_aborted` names from now on, with the reserved objects’ stopping status. These events will be set when the stopping/aborting of the reserved objects had finished. This event flag will be added as a member to `MacroExecutor` which, as on today, is the one in charge of stopping/aborting objects.

There are two design decisions to be taken.

First, when should we return the control to the run macro method?

**Option 1**
We expose the `reserved_objs_stopped` and `reserved_objs_aborted` event flags as the API to the user. Macro run method could wait on this event e.g. in the except (`StopException` or `AbortException`) or in the finally clause. 

**Option 2**
We keep the `reserved_objs_stopped` and `reserved_objs_aborted` event flags protected in the `MacroExecutor` scope. We wait with raising `StopException`, `AbortException` until the reserved objects were stopped/aborted. In this case the synchronization should be done on the `mAPI` decorator level. This way the user will always recover the control in the run method when the objects were already stopped/aborted.

Second, when to execute the `on_stop` and `on_abort` methods?

**Option 1**
We leave to the macro developer the possibility to wait for these event flag somewhere in the `on_stop` or `on_abort` method.

**Option 2**
We postpone execution of `on_stop` or `on_abort` methods until the reserved objects were stopped by waiting on the event flag. In this case, the synchronization should be done on the `MacroExecutor.runMacro` level.

The selected option will be a tradeoff between the macro development API simplicity and the possibility to execute concurrent code in case of macro stop/abort.

In any case, `MacroExecutor` should never consider the macro execution as finished until the reserved objects were stopped/aborted. So the synchronization should be done at some point of the runMacro method.

What is your opinion? Personally I would choose the simplicity.

**Comment by teresanunez** [comment]
I would also use the simplicity, well, and transparency for the
 macro developer.

**Comment by guifrecuni** [comment]
I'd also vote for simplicity in both design decisions.
